### ISSUE TITLE
test stubs placed in wrong target directory?

### ISSUE THREAD

**Issue by boekhold** [groovy/GMavenPlus#52]
Hi,

Cut-down pom.xml:

```
    <properties>
        <maven.compiler.plugin.version>3.5.1</maven.compiler.plugin.version>
        <build.helper.plugin.version>1.10</build.helper.plugin.version>
        <gmavenplus.version>1.5</gmavenplus.version>
    </properties>

   <build>
        <plugins>
            <plugin>
                <groupId>org.codehaus.gmavenplus</groupId>
                <artifactId>gmavenplus-plugin</artifactId>
                <version>${gmavenplus.version}</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>addSources</goal>
                            <goal>addTestSources</goal>
                            <goal>generateStubs</goal>
                            <goal>compile</goal>
                            <goal>testGenerateStubs</goal>
                            <goal>testCompile</goal>
                            <goal>removeStubs</goal>
                            <goal>removeTestStubs</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>

            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>${maven.compiler.plugin.version}</version>
            </plugin>

            <!-- complex looking, but all it does is to make it clear
                 to IntelliJ that the src/main/groovy and src/test/groovy
                 directories contain sources and test sources. If you do not
                 have this, IntelliJ will not treat the directories properly.
            -->
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>build-helper-maven-plugin</artifactId>
                <version>${build.helper.plugin.version}</version>
                <executions>
                    <execution>
                        <id>add-groovy-sources</id>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>add-source</goal>
                        </goals>
                        <configuration>
                            <sources>
                                <source>src/main/groovy</source>
                            </sources>
                        </configuration>
                    </execution>
                    <execution>
                        <id>add-groovy-test-sources</id>
                        <phase>generate-test-sources</phase>
                        <goals>
                            <goal>add-test-source</goal>
                        </goals>
                        <configuration>
                            <sources>
                                <source>src/test/groovy</source>
                            </sources>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```
When I run `mvn test` from inside IntelliJ (2016.1), my test cases from under `src/test/groovy` are being compiled to stubs in `target/generated-sources/groovy-stubs`, and those stubs are not cleaned after the test cases have been run. Because of that IntelliJ gives me a warning on my test classes that "Class XXXX already exists in x.y.z".

What am I missing here?

**Comment by boekhold** [comment]
Ah, if I remove the `testGenerateStubs` and `removeTestStubs` goals, it works. Since I'm not calling groovy test code from java test code, I don't think I need those stubs anyways.

Still, the behavior I saw _with_ those goals seems wrong to me. Java main code would not depend on Groovy test code I'd say, so those stubs should not be placed in `target/generated-sources/groovy-stubs` but in `target/generated-test-sources/groovy-stubs`, no?

**Comment by keeganwitt** [comment]
I'm not sure how standard `generated-test-sources` is (some plugins seem to use it, but some don't).  GMavenPlus follows the pattern GMaven did (stubs for main sources go in `${project.build.directory}/generated-sources/groovy-stubs/main`, and stubs for test sources go in `${project.build.directory}/generated-sources/groovy-stubs/test`).  If you know of anywhere mentioning `generated-test-sources` as a standard directory, I'd appreciate you sharing it.

**Comment by boekhold** [comment]
Hi @keeganwitt, I think this is an IntelliJ-specific issue. Maven certainly has no issues. I've found a few other references to this error as well, and they are all Groovy and/or Grails related:

- http://stackoverflow.com/questions/26987237/idea-groovy-test-class-already-exists
- http://www.thedroidsonroids.com/blog/android/how-to-generate-java-sources-using-buildsrc-gradle-project/ (see "Sample Project" section at the end)

There's an IntelliJ bug report as well that seems related:
- https://youtrack.jetbrains.com/issue/IDEA-121193

Looking at my Project Structure, I can see that the `groovy-stubs/test` directory is actually treated as a Sources folder, not as a Tests folder, see attached screenshot. Not sure if GMavenPlus is providing those instructions to IntelliJ or if this is some heuristics embedded in IntelliJ?
![2016-03-29 09_09_13-project structure](https://cloud.githubusercontent.com/assets/3764084/14098531/f6255d4e-f58d-11e5-8ff4-ee81cad237ac.png)

**Comment by boekhold** [comment]
I opened https://youtrack.jetbrains.com/issue/IDEA-153779 for this. Closing this on the GMavenPlus side.

**Comment by drdamour** [comment]
this has been plaguing me an my team for a while.  we DO have java tests that depend on groovy test sources.  Would it be possible to allow us to override the output location of test generation stubs...if i switch manually in my project the path target/generated-sources/groovy-stubs/tests to be a test source
<img width="443" alt="screen shot 2016-05-15 at 11 12 57 am" src="https://cloud.githubusercontent.com/assets/1514496/15275633/293671f8-1a8e-11e6-81fb-b7096c332fcb.png">
the problem goes away completely.

as such i'd like to re-open this ticket with such a feature request.

**Comment by keeganwitt** [comment]
@drdamour, you can override the default location today. Just set `<testStubsOutputDirectory>` to where you want the stubs written. Unless I'm misunderstanding the question...

**Comment by drdamour** [comment]
ah great!  tried it out

```
        <configuration>
          <testStubsOutputDirectory>${project.build.directory}/generated-test-sources/groovy-stubs/test</testStubsOutputDirectory>
        </configuration>
```

and sure enough it makes intellij detect the stubs as test sources and solved the issue...pretty good workaround i think. 

thanks!

**Comment by drdamour** [comment]
did testStubsOutputDirectory get removed or renamed from 1.5 to 1.6.1?  intellij is complaining to me

**Comment by keeganwitt** [comment]
@drdamour Yes, it looks like that accidentally got changed (it's now just `outputDirectory`).  I have added it to the list of breaking changes for the 1.6 [release notes](https://github.com/groovy/GMavenPlus/releases/tag/1.6).  This is potentially kinda painful to work around, depending on your configuration needs (though it should be possible with multiple executions), so I will be fixing this ASAP.  I have created #120 to track this.  I apologize for the goof, and thank you for reporting this.
As a heads up, the fix will be a breaking change, since it will mean renaming existing configuration properties to avoid conflicts.
